src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -325,6 +330,131 @@ EnsureCitusTableCanBeCreated(Oid relationOid)
+ * udistributes the table.
+ */
+Datum
+undistribute_table(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+	EnsureCoordinator();
+
+	Oid relationId = PG_GETARG_OID(0);
+
+	if (!IsCitusTable(relationId))
+	{
+		ereport(ERROR, (errmsg("Cannot undistribute relation."),
+						errdetail("The relation is not distributed.")));
+	}
+
+	EnsureTableOwner(relationId);
+	Relation relation = try_relation_open(relationId, ExclusiveLock);
+
+	if (relation == NULL)
+	{
+		ereport(ERROR, (errmsg("Cannot undistribute relation"),
+						errdetail("No such distributed relation exists. "
+								  "Might have already been undistributed.")));
+	}
+
+	char *relationName = get_rel_name(relationId);
+
+	Oid schemaId = get_rel_namespace(relationId);
+	char *schemaName = get_namespace_name(schemaId);
+
+
+	if (TableReferencing(relationId))
+	{
+		ereport(ERROR, (errmsg("Cannot undistribute table "
+							   "because it has a foreign key.")));
+	}
+
+	if (TableReferenced(relationId))
+	{
+		ereport(ERROR, (errmsg("Cannot undistribute table "
+							   "because a foreign key references to it.")));
+	}
+
+	if (PartitionedTable(relationId))
+	{
+		ereport(ERROR, (errmsg("Cannot undistribute partitioned tables.")));
+	}
+
+	StringInfo query = makeStringInfo();
+
+	int spiConnectionResult = SPI_connect();
+	if (spiConnectionResult != SPI_OK_CONNECT)
+	{
+		ereport(ERROR, (errmsg("could not connect to SPI manager")));
+	}
+
+	List *tableConstructionCommands = GetTableConstructionCommands(relationId);
+	List *tableCreationCommands = GetTableCreationCommands(relationId, true);
+
+	char *tempName = pstrdup(relationName);
+	uint32 hashOfName = hash_any((unsigned char *) tempName, strlen(tempName));
+	AppendShardIdToName(&tempName, hashOfName);
+
+	char *tableCreationQuery = NULL;
+	foreach_ptr(tableCreationQuery, tableCreationCommands)
+	{
+		Node *parseTree = ParseTreeNode(tableCreationQuery);
+
+		if (nodeTag(parseTree) == T_CreateStmt ||
+			nodeTag(parseTree) == T_AlterOwnerStmt ||
+			nodeTag(parseTree) == T_AlterTableStmt)
+		{
+			RelayEventExtendNames(parseTree, schemaName, hashOfName);
+			CitusProcessUtility(parseTree, tableCreationQuery, PROCESS_UTILITY_TOPLEVEL,
+								NULL, None_Receiver, NULL);
+		}
+	}
+
+	resetStringInfo(query);
+	appendStringInfo(query, "INSERT INTO %s SELECT * FROM %s",
+					 quote_qualified_identifier(schemaName, tempName),
+					 quote_qualified_identifier(schemaName, relationName));
+	SPI_execute(query->data, false, 0);
+
+	List *ownedSequences = getOwnedSequences(relationId, InvalidAttrNumber);
+
+	Oid sequenceOid = InvalidOid;
+	Oid newRelationId = get_relname_relid(tempName, schemaId);
+	foreach_oid(sequenceOid, ownedSequences)
+	{
+		changeDependencyFor(RelationRelationId, sequenceOid,
+							RelationRelationId, relationId, newRelationId);
+	}
+
+	relation_close(relation, NoLock);
+
+	resetStringInfo(query);
+	appendStringInfo(query, "DROP TABLE %s",
+					 quote_qualified_identifier(schemaName, relationName));
```

Comment:
Can there be cases in which the DROP TABLE would fail without a CASCADE? (not sure)

If we add a cascade, can there be cases in which we drop data that the user did not intend to drop?